man/8/qmail-qstat.8
.TH qmail-qstat 8 "" eQmail
.SH NAME
qmail-qstat \- summarize status of mail queue
.SH SYNOPSIS
.B qmail-qstat \fI[options]\fR
.SH DESCRIPTION
\fBqmail-qstat\fR gives a human-readable breakdown of the number of messages at
various spots in the mail queue. \fBqmail-qstat\fR can give more detailed info
about messages in the queue and chn remove messages from the queue.

\fBqmail-qstat\fR must be run either as \fBroot\fR or with group id \fBqmail\fR.
.SH OPTIONS
.TP
-D <id>
removes message \fIid\fR from queue
.TP
-d
deletes all messages from the queue
.TP
-h
shows help and exists
.TP
-i
lists the \fIid\fR's of all message in the queue
.TP
-l
lists all message in the queue
.TP
-L
lists number of messages in the queue with local recipient
.TP
-m <id>
shows the message with \fIid\fR
.TP
-R
lists number of messages in the queue with remote recipient
.TP
Without an option \fBqmail-qstat\fR shows a summary and exists.
.SH "SEE ALSO"
qmail-qread(8),qmail-send(8)


CHANGES
commit 5d9fb974fa147f29fe7d8fc438ffe9cfb70d803d
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 15 16:09:15 2017 +0100

    removed subgetopt/sgetopt

commit cabc31d1eae1b9ffc056142afa9f58ad3e1e2ecf
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 15 13:58:54 2017 +0100

    reworked 'configure' script

commit 6f57aeea75aa12e8fca937de3761d0b2bf0c5d07
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 18:10:04 2017 +0100

    qmail-inject: switched to buffer

commit e3930d23f3df952b6b1276cd7e23722c9dac793f
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 16:59:45 2017 +0100

    qmail-pw2u: switched to 'buffer'

commit bdc2690aecb77ba57aabaf4b4c399ff36cef9789
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 16:45:29 2017 +0100

    tcp-env: switched to getoptb (buffer)

commit e9d22dd4fff1b4fe8395da4d41862e78036ee53f
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 15:36:56 2017 +0100

    preline: switched to 'buffer'

commit ff259662a326a975e56aab07fceae08d9e006298
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 11:25:55 2017 +0100

    added 'socket' lib from ucspi-tcp-088

commit 900d1d7b469ccb793b0b254965523c1f7f3c8a8b
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 14 08:22:24 2017 +0100

    reworked stralloc lib

commit 71e628955516f4055466902a8e1ca100f56f7ba2
Author: Kai Peter <kp@openqmail.org>
Date:   Mon Mar 13 15:40:54 2017 +0100

    improved build/install system, code clean-up

commit 176589f5c769cc60e9321f17ca3cd294a9f067c0
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Mar 10 11:19:09 2017 +0100

    reworked env lib

commit a0b69812ab0e25089f48b21bc69e5dc5597b0a9f
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Mar 9 14:12:38 2017 +0100

    added lib getln_buf

commit 649080aedb1a706827cb1dc935626516cbca9d9e
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Mar 9 14:11:39 2017 +0100

    install target in Makefile

commit 4a591da4d737131fcc9d10458fed0cbfd9af0c17
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 8 16:31:16 2017 +0100

    added fuction 'scan_xlong' to scan lib

commit ed716f2d7587762ed6b1227700e430825f764e03
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 8 16:19:04 2017 +0100

    switched to newer sig() lib from ucspi-tcp-0.88

commit d5b1a2055373b39aa59273a74a775a923c120a37
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 8 16:06:12 2017 +0100

    added tai, taia and iopauce from ucspi-tcp-0.88

commit 205fbd816da6b4098c748e9b293a1337188acc55
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 8 10:09:59 2017 +0100

    replaced vfork() by fork()

commit 749e25226bf845a98e87217398ec005ed0ccd983
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Mar 8 09:12:24 2017 +0100

    updated to tls-patch 20160918

commit 6d1a4dbf3da483e0e9c4beab21538bb75ddedee0
Author: Kai Peter <kp@openqmail.org>
Date:   Mon Mar 6 16:49:27 2017 +0100

    switch to new(er) cdb lib

commit e4fa52e96d947e02d76d11106a8e614d91d3f28f
Author: Kai Peter <kp@openqmail.org>
Date:   Mon Mar 6 11:41:55 2017 +0100

    qmail-local: switched to 'getoptb' and 'buffer'

commit 528df12a41c5156c1f05ad83749b0543389a2de5
Author: Kai Peter <kp@openqmail.org>
Date:   Mon Mar 6 09:38:42 2017 +0100

    added 'getoptb'

commit 83e8b06c4bb1aa8dcf1c54f787bea79b92ee0153
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Mar 3 12:18:21 2017 +0100

    added mkrsadhkeys mksrvrcerts

commit 6f739706612ca4d52b6fa4c4fdefe22ed07397c6
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Mar 3 12:06:49 2017 +0100

    re-worked fmtqfn.{h,c}

commit 96552e68348125c1f3c1f401780a67b703f8d422
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Mar 3 11:19:48 2017 +0100

    added 'buffer' lib

commit 19c9c42879ca13d843640dc935f76d3712153c27
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Mar 2 16:48:22 2017 +0100

    added qmail-shcfg, qmail-fixq

commit 914ea8122b3591adf48f825eb7f9a6ab9db6690e
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Mar 2 16:46:48 2017 +0100

    replaced qmail-qstat

commit e5ba1ad6d98f8a9c51c335a220d4c9d0b2a1ad2a
Author: Kai Peter <kp@openqmail.org>
Date:   Thu Mar 2 16:20:33 2017 +0100

    replaced outdated install script after rollback

commit 576e6e365f60931ef2276c9b5b1458843e441269
Merge: 446793b e595a09
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 14:49:36 2016 +0200

    fixed conflicts

commit e595a096efbdd0317e052b57e3c48df8daa51345
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 14:36:37 2016 +0200

    add files before solve conflicts

commit fc00052ccc96e45be2699bf4fb9fcd224a2b828a
Merge: a76e297 416edce
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 14:34:46 2016 +0200

    Merge branch 'master' into dev

commit 446793ba603e3fc97a3c460efe2e027809035a23
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 12:14:51 2016 +0200

    resolve conflicts

commit b651d37a53bde2074cf31020f224658d964e4868
Merge: 360e06b f643417
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 12:08:28 2016 +0200

    resolve conflicts

commit 360e06b890427969757aa403fc82f4fb947cd58a
Author: Kai Peter <kp@openqmail.org>
Date:   Wed Apr 27 11:35:14 2016 +0200

    1.09 final

commit a76e297b12ebdf5967e264d36691cc1774c635de
Author: Kai Peter <kp@openqmail.org>
Date:   Fri Apr 8 13:11:17 2016 +0200

    1.10p3

commit f64341737d56aa05fca8695e71f658c6589a6d82
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 29 17:00:52 2016 +0200

    Update CHANGLOG-1.09

commit 57b16e354477498fc3f6c443bbbfd6b146b9c99c
Merge: 416edce e4c20f9
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 29 16:59:32 2016 +0200

    Merge pull request #2 from kp-org/dev

    eQmail-1.09 RC2

commit e4c20f947c4251e23bf619a2158e4d552fbe3165
Author: Kai Peter <kp@openqmail.org>
Date:   Tue Mar 29 16:39:57 2016 +0200

    eQmail-1.09 RC2

commit 8a1e087026b1ca66f0bcccff7214833d0bd1aba1
Author: Kai Peter <kp@openqmail.org>
Date:   Sun Mar 27 01:54:09 2016 +0100

    1.10p2

commit cfc7ed9f30bdc01ddc81403c629979ad96562d6f

qmail-qstat.sh
QMAILDIR="QPRFX"
cd $QMAILDIR
DIRS=`ls var/queue/mess/ | wc -l | xargs`   # check and count queue split
# simple try to check out the 'service start' process
for s in eqmaild qmail svscan    # xinetd ???
do
  [ -f "/etc/init.d/$s" ] && BC="$s" && break
done

showHelp() {
  echo "Usage: qmail-qstat [option [value]]"
  echo
  echo -e "  -h\t\tShow this help"
  echo -e "  -D <id>\tDelete message with id 'id' from queue"
  echo -e "  -d\t\tdelete all messages from queue"
  echo -e "  -i\t\tList ID's of all messages in queue"
  echo -e "  -l\t\tList all messages in queue"
  echo -e "  -L\t\tCount messages with local receipient"
  echo -e "  -m <id>\tShow message with id 'id'"
  echo -e "  -R\t\tCount messages with remote receipient"
  echo
  exit 0
}

countBounces() {
  bouncefiles=`ls var/queue/bounce | wc -w | xargs`
  echo "           messages with bounces: "$bouncefiles
}
countLocal() {
  localfiles=`find var/queue/local/* -print | wc -w`
  echo " messages with local receipients: "`expr $localfiles - $DIRS`
}
countRemote() {
  remotefiles=`find var/queue/remote/* -print | wc -w`
  echo "messages with remote receipients: "`expr $remotefiles - $DIRS`
}
countTodo() {
  todofiles=`find var/queue/todo/* -print | wc -w`
  echo "  messages not pre-processed yet: "`expr $todofiles - $DIRS`
}
default() {
  echo -e "\nSummary:"
  messfiles=`find var/queue/mess/* -print | wc -w`
  echo "         total messages in queue: "`expr $messfiles - $DIRS`
  countLocal
  countRemote
  countTodo
  countBounces
  echo -e "\nDirectory split of the queue is $DIRS.\n"
}
listMessages() {
  [ "$QSUBDIR" ] || QSUBDIR="mess"
  [ "$N" ] && FARGS="-name $N"
  for M in $(find var/queue/$QSUBDIR/*/* $FARGS -print 2>/dev/null)
  do echo -e "\nMessage-Id: `basename $M`" ; cat $M ; done
}
listMsgIDs() {
  echo "ID's of messages in queue:"
  [ "$QSUBDIR" ] || QSUBDIR="mess"
  for ID in $(find var/queue/$QSUBDIR/*/* -print 2>/dev/null)
  do echo `basename "$ID"` ; done
}
stopServices() {
  # return and continue w/o stopping services
  [ ! "$BC" ] && echo "Warning: couldn't stop services ..." && return
  echo "stopping services ..."
  # try to stop service and if it fails - bummer!
  /etc/init.d/"$BC" stop 2>&1 >/dev/null
}
startServices() {
  [ ! "$BC" ] && echo "Please restart 'qmail-send' now!" && return
  echo "restarting services ..."
  #
  /etc/init.d/"$BC" start #2>&1 >/dev/null
}
deleteMessage() {
  stopServices
  [ "$QSUBDIR" ] || QSUBDIR="mess"
  [ "$N" ] && FARGS="-name $N"	# set 'find' arguments for a single message
  echo -n "deleting message"
  [ "$N" ] && echo -n " $N" || echo -n "s"	# be a bit verbose ...
  # find and delete message(s)
  for M in $(find var/queue/*/*/* $FARGS 2>/dev/null | xargs)
  do
    Q=`echo $M | cut -d/ -f2`
    rm -f "$M"
  done
  if [ "$M" ] ; then echo "... done" ;		# print result
    else [ "$FARGS" ] && \
      echo " ... failed: no such message!" || \
      echo " ... no messages in queue found!"
  fi
  startServices
}

if [ ! $1 ] ; then default ; exit 0 ; fi
while getopts ":hliLRm:D:d" o; do
    case "${o}" in
        L) QSUBDIR="local" ; listMessages ; echo ; countLocal ; echo ; exit;;
        R) QSUBDIR="remote" ; listMessages ; echo ; countRemote ; echo ; exit;;
        h) showHelp;;
        i) listMsgIDs;;
        l) listMessages ; default;;
        m) N=${OPTARG} ; listMessages ; echo;;
        D) N=${OPTARG} ; deleteMessage ;;
        d) deleteMessage;;

        *) echo "Invalid option!" ; showHelp;;
    esac
done ; shift $((OPTIND-1))

exit 0
